core/tx_list.go

```diff
@@ -521,7 +521,11 @@ func (l *txPricedList) Removed(count int) {
 // Underpriced checks whether a transaction is cheaper than (or as cheap as) the
 // lowest priced (remote) transaction currently being tracked.
 func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
-	return l.underpricedFor(&l.urgent, tx) && l.underpricedFor(&l.floating, tx)
+	// Note: with two queues, being underpriced is defined as being worse than the worst item
+	// in all non-empty queues if there is any. If both queues are empty then nothing is underpriced.
+	return (l.underpricedFor(&l.urgent, tx) || len(l.urgent.list) == 0) &&
```

Comment:
I don't get these special cases. Why return "underpriced" if a queue is empty?

Reply:
See the comment above. We don't report underpriced if a queue is empty, we just don't automatically consider it non-underpriced. The last line fixes the condition that I initially broke, so if there are no non-empty queues then obviously nothing is underpriced.

Reply:
I still don't get it why we need to care if a queue is empty. If it's empty, there's room to put the tx in there, so it's not underpriced. What am I missing?

Reply:
It's just a corner case that's not going to happen normally but for example a test fails if I don't have this check because it uses a low limit and nothing fits in the floating queue which would make nothing underpriced, ever. I think this condition is good enough for those corner cases while in normal case where 1/4 of the total remotes count is always >= 1 it doesn't matter.